### PR TITLE
Improve AMP sanitization for title tags and meta descriptions

### LINKED ISSUE #12655: Web Stories plugin not printing title tag by default
## Bug Description
Stories are missing the <title> tag when using a theme like Astra (astra) Version 3.9.3 or Twenty Twenty Version 2.1.

Example Story URL: https://testsite.cmsdevrel.com/web-stories/testing-nov-9-5

## Expected Behaviour

Web Stories plugin prints title tags by default.

## Steps to Reproduce

1. Activate a theme like Astra or Twenty Twenty
2. View a single story
3. Verify there is no  <title> tag in the source code

## Additional Context

- Plugin Version: 1.27.0
- WordPress Version:
- Operating System: macOS
- Browser: Chrome

### LINKED ISSUE #12709: Code quality: `Use str_starts_with` and `str_ends_with` in PHP code
## Feature Description

Use `Use str_starts_with` and `str_ends_with` in PHP code as pollyfill were added in WordPress 5.9

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

We had issues with missing title tags / meta descriptions or duplicate ones, depending on the theme and SEO plugin being used.

I also contributed a [PR to Yoast SEO](https://github.com/Yoast/wordpress-seo/pull/19214) to improve their logic, as it turns out they did some odd changes there recently.

## Summary

Prevent missing title tags / meta descriptions or duplicate ones at the AMP sanitization level.

Also added another fix to ensure links always have a proper protocol.

## User-facing changes

No visual changes

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. View a published story on the frontend
2. Verify that there's a `<title>` tag and a meta description present

Test with:

* Twenty Twelve theme
* Twenty Twenty-Three theme
* OceanWP theme
* Yoast SEO enabled/disabled

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12655
Fixes #12709

### CODE REVIEW

File: includes/AMP/Sanitization.php [hunk 1]
```diff
@@ -442,6 +442,8 @@ protected function get_sanitizers(): array {
 				'publisher'      => $story->get_publisher_name(),
 				'poster_images'  => array_filter( $poster_images ),
 				'video_cache'    => $video_cache_enabled,
+				'title_tag'      => wp_get_document_title(),
+				'description'    => wp_strip_all_tags( get_the_excerpt() ),
```
Comment (spacedmonkey) [inline includes/AMP/Sanitization.php:446]: What about excerpt length and `wp_trim_excerpt`? Adding `[&hellip;]` at the end here might not be the best thing.

Comment (spacedmonkey) [reply includes/AMP/Sanitization.php:446]: Why not use `$story->get_excerpt()`?

File: includes/AMP/Traits/Sanitization_Utils.php [hunk 2]
```diff
@@ -113,6 +115,17 @@ private function transform_a_tags( $document ): void {
...
 				$link->removeAttribute( 'data-tooltip-text' );
 			}
+
+			// Extra hardening to catch links without a proper protocol.
+			// Matches withProtocol() util in the editor.
+			if (
+				0 !== stripos( $url, 'http://' ) &&
+				0 !== stripos( $url, 'https://' ) &&
+				0 !== stripos( $url, 'tel:' ) &&
+				0 !== stripos( $url, 'mailto:' )
```
Comment (spacedmonkey) [inline includes/AMP/Traits/Sanitization_Utils.php]: Could we use `wp_allowed_protocols()` here?

Comment (spacedmonkey) [reply includes/AMP/Traits/Sanitization_Utils.php]: Maybe look at `wp_kses_bad_protocol_once`.

File: includes/AMP/Traits/Sanitization_Utils.php [hunk 3]
```diff
@@ -613,4 +626,83 @@ private function remove_page_template_placeholder_images( $document ): void {
...
+		$meta_descriptions = $document->xpath->query( './/meta[@name="description"]' );
+
+		if ( $meta_descriptions->length > 1 ) {
+			foreach ( $meta_descriptions as $index => $meta_description ) {
+				if ( 0 === $index ) {
+					continue;
+				}
+				$document->head->removeChild( $meta_description );
+			}
+		}
```
Comment (spacedmonkey) [inline includes/AMP/Traits/Sanitization_Utils.php:693]: These lines are repeated from above. https://github.com/GoogleForCreators/web-stories-wp/pull/12706/files#diff-56d46ddccbd78e7364b5120cf250aff849e7af145759193537cfcb8c3c27fdfeR651-R658. Move this logic to a method.

File: includes/Integrations/AMP.php [hunk 4]
```diff
@@ -234,6 +234,8 @@ public function add_amp_content_sanitizers( $sanitizers ) {
 			'publisher'      => $story->get_publisher_name(),
 			'poster_images'  => array_filter( $poster_images ),
 			'video_cache'    => $video_cache_enabled,
+			'title_tag'      => wp_get_document_title(),
+			'description'    => wp_strip_all_tags( get_the_excerpt() ),
```
Comment (spacedmonkey) [inline includes/Integrations/AMP.php:238]: See https://github.com/GoogleForCreators/web-stories-wp/pull/12706/files#r1026440425

File: includes/AMP/Sanitization.php (hunk 1 above)
Comment (swissspidy) [reply includes/AMP/Sanitization.php:446]: This is copied from the Discovery class. Has the advantage that it has a maximum length and is filterable.

File: includes/AMP/Traits/Sanitization_Utils.php (hunk 2 above)
Comment (swissspidy) [reply includes/AMP/Traits/Sanitization_Utils.php]: This matches the list of allowed protocols in the editor, as per the inline comment.

File: includes/AMP/Traits/Sanitization_Utils.php (hunk 3 above)
Comment (swissspidy) [reply includes/AMP/Traits/Sanitization_Utils.php:693]: A bit overkill IMO to add an extra method for this

File: includes/AMP/Sanitization.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/AMP/Sanitization.php:446]: Is clipping the text a good thing in this place?

File: includes/AMP/Traits/Sanitization_Utils.php (hunk 2 above)
Comment (spacedmonkey) [reply includes/AMP/Traits/Sanitization_Utils.php]: Okay, how about an array you loop around then?

File: includes/AMP/Sanitization.php (hunk 1 above)
Comment (swissspidy) [reply includes/AMP/Sanitization.php:446]: It's the existing behavior of the plugin.

I don't see an issue with clipping. While there's no limit on how long a meta description can be, the snippet is typically truncated in Google Search results as needed anyway.

File: includes/AMP/Traits/Sanitization_Utils.php [hunk 5]
```diff
@@ -88,7 +88,9 @@ private function transform_a_tags( $document ): void {
 				$link->setAttribute( 'target', '_blank' );
 			}
 
-			$is_link_to_same_origin = 0 === strpos( $link->getAttribute( 'href' ), home_url() );
+			$url = $link->getAttribute( 'href' );
+
+			$is_link_to_same_origin = 0 === strpos( $url, home_url() );
```
Comment (spacedmonkey) [inline includes/AMP/Traits/Sanitization_Utils.php]: How about using `str_starts_with`? There is a polyfill in core for this.

File: includes/AMP/Traits/Sanitization_Utils.php (hunk 2 above)
Comment (spacedmonkey) [reply includes/AMP/Traits/Sanitization_Utils.php]: Again we can use `str_starts_with` here.

File: includes/AMP/Traits/Sanitization_Utils.php (hunk 3 above)
Comment (spacedmonkey) [reply includes/AMP/Traits/Sanitization_Utils.php:693]: It makes it more testable....

Comment (spacedmonkey) [reply includes/AMP/Traits/Sanitization_Utils.php:693]: And reduces complexity of method.

File: includes/AMP/Traits/Sanitization_Utils.php (hunk 2 above)
Comment (swissspidy) [reply includes/AMP/Traits/Sanitization_Utils.php]: You mean something like this?

```php
$has_protocol = false;
foreach ( [ 'http://', 'https://', 'tel:', 'mailto:' ] as $protocol ) {
	if ( 0 === stripos( $url, $protocol ) ) {
		$has_protocol = true;
		break;
	}
}

if ( ! $has_protocol ) {
	$link->setAttribute( 'href', 'https://' . $url );
}
```

That is much longer and less readable

File: includes/Discovery.php [hunk 6]
```diff
@@ -103,26 +103,25 @@ public function register(): void {
...
-
 		/**
 		 * Filters whether to print the document title.
 		 *
 		 * @since 1.25.0
 		 *
 		 * @param bool $enable_open_graph Whether to print the document title. Defaults to true for block themes, false otherwise.
 		 */
-		$enable_metadata = apply_filters( 'web_stories_enable_document_title', $enable_metadata );
+		$enable_metadata = apply_filters( 'web_stories_enable_document_title', true );
```
Comment (spacedmonkey) [inline includes/Discovery.php:124]: Update the comment above?

File: includes/AMP/Traits/Sanitization_Utils.php (hunk 5 above)
Comment (swissspidy) [reply includes/AMP/Traits/Sanitization_Utils.php]: Oh neat, I didn't know these were added in WordPress 5.9. I thought they were added later.

Makes total sense in that case!

File: includes/AMP/Sanitization.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/AMP/Sanitization.php:446]: Okay, cool.Just checking.

File: includes/AMP/Traits/Sanitization_Utils.php (hunk 2 above)
Comment (spacedmonkey) [reply includes/AMP/Traits/Sanitization_Utils.php]: I find the more readable. But we have had this discussion before..

**Review by spacedmonkey — APPROVED** [review]
I think we are looking good here.

**Comment by kkalarickal** [comment]
Confirmed that the issue was present in `main branch` with Twenty Twelve theme

**Comment by kkalarickal** [comment]
Verified fix in `12706 branch` on stories-qa server

**Comment by kkalarickal** [comment]
Confirmed that the title tag is printed for: Twenty Twenty Three and Astra themes, as well as with Yoast SEO plugin enabled and disabled.

**Comment by kkalarickal** [comment]
@swissspidy - the summary of this ticket talks about 

> Also added another fix to ensure links always have a proper protocol.

Q: How do I test this ? 

I added some products and embedded some links within the page or in images within the page. When I examined the headers I could find the embedded links in the page. Is there anything else I should check for this part of the fix ?

**Comment by swissspidy** [comment]
No need to test that part 👍

**Pull request merged**